File: dt241m-controller/internal/testutil/fakemqtt.go
package testutil

import (
	"context"
	"encoding/json"
	"strings"
	"sync"

	"github.com/jacobgad/dt241m-controller/internal/mqtt"
)

type Published struct {
	Topic   string
	Payload string
	Retain  bool
}

type FakeMQTT struct {
	mu            sync.Mutex
	published     []Published
	subscriptions []string
	onMessage     []mqtt.MessageHandler
	onConnect     []func()
	connected     bool
	Ended         bool
}

func NewFakeMQTT(connected bool) *FakeMQTT {
	return &FakeMQTT{connected: connected}
}

func (f *FakeMQTT) Publish(_ context.Context, topic, payload string, retain bool) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.published = append(f.published, Published{Topic: topic, Payload: payload, Retain: retain})
	return nil
}

func (f *FakeMQTT) Subscribe(_ context.Context, topics []string) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.subscriptions = append(f.subscriptions, topics...)
	return nil
}

func (f *FakeMQTT) OnMessage(h mqtt.MessageHandler) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.onMessage = append(f.onMessage, h)
}

func (f *FakeMQTT) OnConnect(h func()) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.onConnect = append(f.onConnect, h)
}

func (f *FakeMQTT) Connected() bool {
	f.mu.Lock()
	defer f.mu.Unlock()
	return f.connected
}

func (f *FakeMQTT) Close(context.Context) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.Ended = true
	f.connected = false
	return nil
}

func (f *FakeMQTT) Deliver(topic, payload string) {
	f.mu.Lock()
	handlers := append([]mqtt.MessageHandler{}, f.onMessage...)
	f.mu.Unlock()
	for _, h := range handlers {
		h(topic, []byte(payload))
	}
}

func (f *FakeMQTT) SimulateDisconnect() {
	f.mu.Lock()
	f.connected = false
	f.mu.Unlock()
}

func (f *FakeMQTT) SimulateConnect() {
	f.mu.Lock()
	f.connected = true
	handlers := append([]func(){}, f.onConnect...)
	f.mu.Unlock()
	for _, h := range handlers {
		h()
	}
}

func (f *FakeMQTT) Clear() {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.published = nil
}

func (f *FakeMQTT) Published() []Published {
	f.mu.Lock()
	defer f.mu.Unlock()
	return append([]Published(nil), f.published...)
}

func (f *FakeMQTT) Subscriptions() []string {
	f.mu.Lock()
	defer f.mu.Unlock()
	return append([]string(nil), f.subscriptions...)
}

func (f *FakeMQTT) MessagesOn(topic string) []Published {
	var out []Published
	for _, p := range f.Published() {
		if p.Topic == topic {
			out = append(out, p)
		}
	}
	return out
}

func (f *FakeMQTT) PayloadsOn(topic string) []string {
	var out []string
	for _, p := range f.MessagesOn(topic) {
		out = append(out, p.Payload)
	}
	return out
}

func (f *FakeMQTT) LastOn(topic string) (Published, bool) {
	msgs := f.MessagesOn(topic)
	if len(msgs) == 0 {
		return Published{}, false
	}
	return msgs[len(msgs)-1], true
}

func (f *FakeMQTT) LastPayload(topic string) string {
	last, _ := f.LastOn(topic)
	return last.Payload
}

func (f *FakeMQTT) Retained() map[string]string {
	out := make(map[string]string)
	for _, p := range f.Published() {
		if !p.Retain {
			continue
		}
		if p.Payload == "" {
			delete(out, p.Topic)
		} else {
			out[p.Topic] = p.Payload
		}
	}
	return out
}

func (f *FakeMQTT) DiscoveryConfigs() map[string]map[string]any {
	out := make(map[string]map[string]any)
	for topic, payload := range f.Retained() {
		if strings.HasPrefix(topic, "homeassistant/") && strings.HasSuffix(topic, "/config") {
			var cfg map[string]any
			if json.Unmarshal([]byte(payload), &cfg) == nil {
				out[topic] = cfg
			}
		}
	}
	return out
}

File: dt241m-controller/src/tests/helpers/fake-mqtt.ts
import type { MessageHandler, MqttConnection, PublishOptions } from "../../mqtt/client.js";

export type PublishedMessage = { topic: string; payload: string; retain: boolean; qos: 0 | 1 };

export class FakeMqtt implements MqttConnection {
  readonly published: PublishedMessage[] = [];
  readonly subscriptions: string[] = [];
  private readonly messageHandlers: MessageHandler[] = [];
  private readonly connectHandlers: Array<() => void> = [];
  private readonly disconnectHandlers: Array<() => void> = [];
  private connected: boolean;
  ended = false;

  constructor(initiallyConnected = true) {
    this.connected = initiallyConnected;
  }

  async publish(topic: string, payload: string, options: PublishOptions): Promise<void> {
    this.published.push({ topic, payload, retain: options.retain, qos: options.qos ?? 1 });
  }

  async subscribe(topics: readonly string[]): Promise<void> {
    this.subscriptions.push(...topics);
  }

  onMessage(handler: MessageHandler): void {
    this.messageHandlers.push(handler);
  }

  onConnect(handler: () => void): void {
    this.connectHandlers.push(handler);
  }

  onDisconnect(handler: () => void): void {
    this.disconnectHandlers.push(handler);
  }

  isConnected(): boolean {
    return this.connected;
  }

  async end(): Promise<void> {
    this.ended = true;
    this.connected = false;
  }

  deliver(topic: string, payload: string): void {
    for (const handler of this.messageHandlers) handler(topic, payload);
  }

  simulateDisconnect(): void {
    this.connected = false;
    for (const handler of this.disconnectHandlers) handler();
  }

  simulateConnect(): void {
    this.connected = true;
    for (const handler of this.connectHandlers) handler();
  }

  clear(): void {
    this.published.length = 0;
  }

  messagesOn(topic: string): PublishedMessage[] {
    return this.published.filter((m) => m.topic === topic);
  }

  lastOn(topic: string): PublishedMessage | undefined {
    const matches = this.messagesOn(topic);
    return matches[matches.length - 1];
  }

  retainedSnapshot(): Map<string, string> {
    const snapshot = new Map<string, string>();
    for (const message of this.published) {
      if (!message.retain) continue;
      if (message.payload === "") snapshot.delete(message.topic);
      else snapshot.set(message.topic, message.payload);
    }
    return snapshot;
  }

  discoveryConfigs(): Map<string, Record<string, unknown>> {
    const configs = new Map<string, Record<string, unknown>>();
    for (const [topic, payload] of this.retainedSnapshot()) {
      if (topic.startsWith("homeassistant/") && topic.endsWith("/config")) {
        configs.set(topic, JSON.parse(payload) as Record<string, unknown>);
      }
    }
    return configs;
  }
}
